### PR TITLE
Add Module#descendants

### PR DESCRIPTION
**Pull request by fatkodima**
This method allows to return specific class/module descendants.

Relevant discussion - https://bugs.ruby-lang.org/issues/14394

```ruby
 class A
 end
 
 class B < A
 end
 
 class C < B
 end
 
 A.descendants    #=> [A, B, C]
 B.descendants    #=> [B, C]
 C.descendants    #=> [C]
```

### CODE REVIEW

**Comment by Hanmac** [comment]
@fatkodima how does this handle anonymous classes? how about adding a parameter to exclude/include them?

**Comment by simi** [comment]
@Hanmac please can you copy your comment to https://bugs.ruby-lang.org/issues/14394 where this is discussed?

File: class.c [hunk 1]
```diff
@@ -1269,6 +1269,60 @@ rb_mod_ancestors(VALUE mod)
...
 }
 
+static void
+collect_descendants(VALUE mod, VALUE hash)
+{
+    if (BUILTIN_TYPE(mod) != T_ICLASS) {
+        rb_hash_aset(hash, mod, Qnil);
+    }
+
+    rb_class_foreach_subclass(mod, collect_descendants, hash);
```
Comment (eregon) [inline class.c:1279]: Does `rb_class_foreach_subclass` work for modules too?

Comment (fatkodima) [reply class.c:1279]: Yes.
https://github.com/ruby/ruby/pull/3471/files#diff-02bfcb686555945e00ad1d80a1ece424R225-R233

**Comment by ioquatix** [comment]
All tests passing, this LGTM.

**Comment by ioquatix** [comment]
@fatkodima can you please squash into one commit if you have time? Thanks!

**Comment by eregon** [comment]
@fatkodima It would be great if you can add some specs in file `spec/ruby/core/module/descendants_spec.rb` and test more corner cases, like `someClass.descendants` should not includes modules, whether the result should include `self`, etc.

File: spec/ruby/core/module/descendants_spec.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      assert_descendants(a, [a, b, c, klass1, klass2])
+    end
+  end
+
+  it "has 1 entry per module or class" do
+    ModuleSpecs::Parent.descendants.should == ModuleSpecs::Parent.descendants.uniq
+  end
+
+  def assert_descendants(mod, descendants)
+    mod.descendants.map(&:name).sort.should == descendants.map(&:name).sort
```
Comment (eregon) [inline spec/ruby/core/module/descendants_spec.rb:50]: `Module.new.name` is `nil`, so better not do the `map(&:name)` (`Module#inspect` shows the name is named)
